### PR TITLE
add JSON IO based on Circe #570

### PR DESCRIPTION
**Pull request by nevillelyh**

### CODE REVIEW

**Comment by nevillelyh** [comment]
@ravwojdyla @regadas thoughts?

File: scio-extra/src/main/scala/com/spotify/scio/extra/json/package.scala [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+  implicit class JsonSCollection[T: ClassTag : Encoder : Decoder](val self: SCollection[T]) {
+    def saveAsJsonFile(path: String, numShards: Int = 0): Future[Tap[T]] = {
+      if (self.context.isTest) {
+        self.context.testOut(JsonIO[T](path))(self)
+        self.saveAsInMemoryTap
+      } else {
+        val encoder = implicitly[Encoder[T]]
+        val decoder = implicitly[Decoder[T]]
+        self
+          .map(_.asJson(encoder).noSpaces)
```
Comment (ravwojdyla) [inline scio-extra/src/main/scala/com/spotify/scio/extra/json/package.scala]: Do you need to specify those explicitly, the seem to be in the context, and `asJson` takes implicit encoder?

Comment (nevillelyh) [inline scio-extra/src/main/scala/com/spotify/scio/extra/json/package.scala]: Yeah there's some issues, fixing.

File: scio-extra/src/main/scala/com/spotify/scio/extra/json/package.scala [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+  /**
+   * Enhanced version of [[SCollection]] with JSON methods.
+   */
+  implicit class JsonSCollection[T: ClassTag : Encoder : Decoder](val self: SCollection[T]) {
+    def saveAsJsonFile(path: String, numShards: Int = 0): Future[Tap[T]] = {
+      if (self.context.isTest) {
+        self.context.testOut(JsonIO[T](path))(self)
+        self.saveAsInMemoryTap
+      } else {
+        val encoder = implicitly[Encoder[T]]
```
Comment (ravwojdyla) [inline scio-extra/src/main/scala/com/spotify/scio/extra/json/package.scala]: Do you need those in question below ↓

File: scio-extra/src/main/scala/com/spotify/scio/extra/json/package.scala [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+        val decoder = implicitly[Decoder[T]]
+        self
+          .map(_.asJson(encoder).noSpaces)
+          .applyInternal(self.textOut(path, ".json", numShards))
+        self.context.makeFuture(TextTap(path + "/part-*").map(decode[T](_)(decoder).right.get))
```
Comment (ravwojdyla) [inline scio-extra/src/main/scala/com/spotify/scio/extra/json/package.scala]: Same as question above for explicit decoder ↑

File: scio-extra/src/main/scala/com/spotify/scio/extra/json/package.scala (hunk 1 above)
Comment (nevillelyh) [inline scio-extra/src/main/scala/com/spotify/scio/extra/json/package.scala]: Actually I need these to defeat closure, will add comments once I fix all the ser/de issues :)

Comment (ravwojdyla) [inline scio-extra/src/main/scala/com/spotify/scio/extra/json/package.scala]: Mark `self` as `@transient` and make `JsonSCollection` extend `Serializable` and you should be good.

Comment (nevillelyh) [inline scio-extra/src/main/scala/com/spotify/scio/extra/json/package.scala]: That works. But it also means `JsonSCollection` gets pulled into the closure. I feel `val encoder = ...` is more explicit w.r.t. what gets pulled in. Thoughts?

Comment (ravwojdyla) [inline scio-extra/src/main/scala/com/spotify/scio/extra/json/package.scala]: What's the problem of `JsonSCollection` getting pulled into the closure. Honestly those `vals` are only explicit if you already know what is going on :)

Comment (nevillelyh) [inline scio-extra/src/main/scala/com/spotify/scio/extra/json/package.scala]: Nothing. But IMO it's better to be clear what gets into the closure (`Encoder`, `Decoder`) and risks. Those implicits could pull in a lot more and all needs to be serializable.

Comment (ravwojdyla) [inline scio-extra/src/main/scala/com/spotify/scio/extra/json/package.scala]: I think the transient/serializable is simpler and more idiomatic solution for the current problem. But it does not matter much cause its not exposed to the user. I will leave it up to you.

Comment (nevillelyh) [inline scio-extra/src/main/scala/com/spotify/scio/extra/json/package.scala]: You convinced me :)
Let's do this. We don't have this problem for any other implicit `sc` or `SColl`s but if we do can apply the same pattern.

**Comment by regadas** [comment]
@nevillelyh @ravwojdyla late to the party, but this looks really nice and clean! well done! :)

**Pull request merged**